src/agent.c

```diff
@@ -547,6 +553,44 @@ int agent_add_remote_candidate(juice_agent_t *agent, const char *sdp) {
 	return ret;
 }
 
+int agent_add_turn_server(juice_agent_t *agent, juice_turn_server_t *turn_server) {
+	int ret = -1;
+	conn_lock(agent);
+
+	if (agent->conn_impl) {
+		JLOG_WARN("Candidates gathering already started");
+		goto exit;
+	}
+
+	JLOG_VERBOSE("adding turn");
+
+	agent->config.turn_servers_count++;
+	if (agent->config.turn_servers_count == 1) {
+		agent->config.turn_servers = calloc(1, sizeof(juice_turn_server_t));
+	} else {
+		agent->config.turn_servers =
+		    realloc(agent->config.turn_servers,
+		            agent->config.turn_servers_count * sizeof(juice_turn_server_t));
+	}
+
+	if (!agent->config.turn_servers) {
+		JLOG_FATAL("Memory allocation for TURN servers copy failed");
+		goto exit;
+	}
+
+	if (!agent_copy_turn_server(&agent->config.turn_servers[agent->config.turn_servers_count - 1],
+	                            turn_server)) {
+		goto exit;
+	}
+
+	ret = 0;
+
+exit:
+	conn_unlock(agent);
+	conn_interrupt(agent);
```

Comment:
You should not call `conn_unlock()` or `conn_interrupt()`, it has no effect as there is no `conn_impl`.